airflow/notifications/basenotifier.py

```diff
@@ -79,15 +79,30 @@ def notify(self, context: Context) -> None:
         """
         ...
 
-    def __call__(self, context: Context) -> None:
+    def __call__(self, *args) -> None:
         """
         Send a notification.
 
         :param context: The airflow context
         """
-        context = self._update_context(context)
-        self.render_template_fields(context)
+        # Currently, there are two ways a callback is invoked
+        # 1. callback(context) - for on_*_callbacks
+        # 2. callback(dag, task_list, blocking_task_list, slas, blocking_tis) - for sla_miss_callback
+        # we have to distinguish between the two calls so that we can prepare the correct context,
+        # comparing len(args) is one straightforward way of checking this.
+        if len(args) == 1:
+            _context = args[0]
```

Comment:
This will still causes issues. What if the first arg is `dag`?.
Suggestion:
Let's have a method that implements sla callback in this base notifier. eg. `sla_notify`. It should take all the necessary args that normal sla callback takes. The notifier `__call__ `should expect an arg called `is_sla` which when true would call the `sla_notify` instead of notify

Reply:
It can be but there will always be other params passed as well - dag, task_list, blocking_task_list, slas and blocking_tis right? 

Since there are only two ways we are calling callback -
1. callback(context) - for on_*_callbacks
2. callback(dag, task_list, blocking_task_list, slas, blocking_tis) - for sla_miss_callback

Reply:
I spent time today with this, here's my suggestion:
Add sla_context in the list of callback args in dag_processing:
`callback(dag, task_list, blocking_task_list, slas, blocking_tis, sla_context=context)`
The `context` should be from the `processor.py` module. There's a function `ti.get_template_context(session=session)` that you can use to get the context you pass to sla_context.
Also add all the args: `dag, task_list, blocking_task_list, slas, blocking_tis` to the context that you got there.
So in base notifier, you do something like:
```python
    def __call__(self, *args, sla_context=None) -> None:
        """
        Send a notification.

        :param context: The airflow context
        """
        context = args if len(args) == 1 else None

        if not context:
            context = sla_context
        ...
```
I think the challenge is getting the ti to use to get the templates but it's doable

Reply:
@ephraimbuddy I think passing extra `sla_context=context` to callback will break the existing sla_miss_callback callbacks. 

```
def test(a, b):
    print("a: ", a, " b: ", b)

test(1, 2, c=3)
```
`Error : TypeError: test() got an unexpected keyword argument 'c'`

Reply:
Yeah. That's true. I wonder how possible it would be to use `provide_session` and get context from one of the dags task instances?

Reply:
Maybe we don't need it? Since `sla_miss_callback` users are expecting parameters - `dag, task_list, blocking_task_list, slas, blocking_tis` it would be should be fine to add those params in context and not the `ti.get_template_context(session=session)` ??

Reply:
@ephraimbuddy Maybe we don't need them? Since the users of `sla_miss_callback` is expecting params `dag, task_list, blocking_task_list, slas, blocking_tis` in their callbacks. It should be fine to create context out of those params and use them in the rendering templates, instead of TI's context. WDYT?

Reply:
What if user decides to use other contexts in the notifier e.g SlackNotifier(text="{{ds}}"), I think it will not work?

Reply:
Ya, it won't work but it will be expected right? Since the value of context changes as per callbacks, you are trying to use it with? 
1. Like on_*_callbacks will have TI's context
2. sla_miss_callback will have `dag, task_list, blocking_task_list, slas, blocking_tis` available in context

We can mention this in the documentation?

Reply:
Yeah makes sense to me. We can mention it in the documentation

Reply:
@ephraimbuddy I have the made change. Please check.